internal/rule/squashjoins.go

```diff
@@ -39,18 +42,103 @@ func SquashJoins(
 
 		return squashJoin(join)
 	})
+
 	if err != nil {
 		return nil, err
 	}
 
-	return n.TransformUp(func(n sql.Node) (sql.Node, error) {
+	n, err = n.TransformUp(func(n sql.Node) (sql.Node, error) {
 		t, ok := n.(*joinedTables)
 		if !ok {
 			return n, nil
 		}
 
 		return buildSquashedTable(t.tables, t.filters, t.columns, t.indexes)
 	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return n.TransformUp(func(n sql.Node) (sql.Node, error) {
+		if projectSquashes <= 0 {
+			return n, nil
+		}
+
+		project, ok := n.(*plan.Project)
+		if !ok {
+			return n, nil
+		}
+
+		child, ok := project.Child.(*plan.Project)
+		if !ok {
+			return n, nil
+		}
+
+		squashedProject, err := squashProjects(project, child)
+		if err != nil {
+			return nil, err
+		}
+
+		projectSquashes--
+		return squashedProject, nil
+	})
+}
+
+func countProjectSquashes(n sql.Node) int {
+	var squashableProjects int
+	plan.Inspect(n, func(node sql.Node) bool {
+		if project, ok := node.(*plan.Project); ok {
+			if _, ok := project.Child.(*plan.InnerJoin); ok {
+				squashableProjects++
+			}
+		}
+
+		return true
+	})
+
+	return squashableProjects - 1
+}
+
+// ErrWrongProjection is raised if a plan.Project node contains a wrong expression.
+var ErrWrongProjection = errors.NewKind("wrong expression found in project node %s")
+
+func squashProjects(parent, child *plan.Project) (sql.Node, error) {
+	projections := []sql.Expression{}
+	for _, expr := range parent.Expressions() {
+		parentField, ok := expr.(*expression.GetField)
```

Comment:
If an expression is a literal, this will fail? Example: 
```
SELECT count(1),1 , refs.repository_id
FROM refs
NATURAL JOIN commits
NATURAL JOIN commit_blobs
WHERE refs.ref_name = 'HEAD'
GROUP BY refs.repository_id
```

Reply:
That query doesn't work because of how the `GroupBy` is implemented right now. It only allows expressions in the select which are aggregations or are in the group by too, not because of the changes in this PR.

The following query, for example. does work:
```
MySQL [(none)]> SELECT 1,refs.repository_id FROM refs NATURAL JOIN commits NATURAL JOIN commit_blobs WHERE refs.ref_name = 'HEAD' limit 10;
+------+---------------+
| 1    | repository_id |
+------+---------------+
|    1 | enry          |
|    1 | enry          |
|    1 | enry          |
|    1 | enry          |
|    1 | enry          |
|    1 | enry          |
|    1 | enry          |
|    1 | enry          |
|    1 | enry          |
|    1 | enry          |
+------+---------------+
10 rows in set (0.02 sec)

```

Reply:
then, how can we reach that error?

Reply:
Sorry @ajnavarro, I think I'm not understanding you. Do you mean that if a literal appears in the projection the query should fail? Anyway that's something not related to this PR, I guess.

Should I open an issue for it?

Reply:
I'm trying to understand when this error will be thrown. 

As I can see on the `squashProjects` you are getting the expressions on the projection and checking if the type is a `Getfield` expression. If not, you are throwing an error.

There are some cases when a projection can contain other expressions than `GetField` (per example, a `Function` or a `Literal`), and that is not necessarily an error.

Reply:
:nerd_face: Now I see what you mean. The thing is that the projections to squash coming from the replacement of natural joins by inner joins, and those projections only contains `GetFields`. That's the reason because the chained projections appear, so functions and literals written in a query aren't going to be squashed.